src/parquet/ReadParquet.cpp

```diff
@@ -1,5 +1,407 @@
 #include "ReadParquet.h"
-#include "UtilParquet.h"
+
+
+namespace akcpp {
+
+namespace {
+
+// ChplBuf::ElemType can be used to get the type of the Chapel buffer from an
+// Arrow type.
+template<typename ArrowType>
+struct ChplBuf {
+  using ElemType = typename ArrowType::c_type;
+};
+
+// specialization for storing 32-bit data in 64-bit arrays
+template<> struct ChplBuf<arrow::Int32Type>  { using ElemType = int64_t; };
+template<> struct ChplBuf<arrow::UInt32Type> { using ElemType = uint64_t; };
+template<> struct ChplBuf<arrow::FloatType>  { using ElemType = double; };
+
+// TODO The implementation seemingly reads 128-bit decimal values into
+// Chapel `real`s (or C `double`s). I don't see how this can work properly with
+// large enough data. For now, I think this keeps the implementation practically
+// same as the old, likely carrying the same bug into the new implementation,
+// unfortunately.
+template<> struct ChplBuf<parquet::FLBAType>  {using ElemType = double;};
+
+// this struct is used to bundle different types into one, based on a single
+// ArrowType
+template<typename ArrowType>
+struct TypeBundle {
+  using ReaderType = typename parquet::TypedColumnReader<ArrowType>;
+  using ChplType = typename ChplBuf<ArrowType>::ElemType;
+  using PqType = typename ArrowType::c_type;
+};
+
+template<>
+struct TypeBundle<arrow::Decimal128Type> {
+  using ReaderType = typename parquet::TypedColumnReader<parquet::FLBAType>;
+  using ChplType = typename ChplBuf<parquet::FLBAType>::ElemType;
+  using PqType = typename parquet::FixedLenByteArray;
+};
+} // end anonymous namespace
+
+// Engin: We could consider exposing this struct to Chapel. Note that in
+// ParquetMsg.chpl there is a somewhat similar type. However, this one is for a
+// single column read
+// Based on what we decide, the implementation of this struct can be mode more
+// C++-like or could stay more C-like.
+struct AkColReadOp {
```

Comment:
nit, this is already inside a namespace called `akcpp`, no reason to further prefix the name with `Ak`

Reply:
That's a really nice nit. Fixed.